# Examples and Exercises from Think Stats, 2nd Edition

http://thinkstats2.com

Copyright 2016 Allen B. Downey

MIT License: https://opensource.org/licenses/MIT


In [28]:
from __future__ import print_function, division
import nsfg

In [52]:
import pandas as pd
import sqlanydb
from string import Template

mydf1 = pd.read_csv('mydata/test.txt')
sql = '''select item, brand, model, qty, price, "date"
    from bill_memo key join bill_memo_product
        key join inv_main
            key join product
                order by "date" '''
def getconn(db):
    return sqlanydb.connect(uid='dba', pwd='sql', eng='server', dbn=db, host='kushserver' )

def get_amount(row):
    a=  (float(row['qty']) * float(row['price']))
    return(a)
    #return ("{:,}".format(a)) 

pd.options.display.float_format = '{:,.0f}'.format

dfa = pd.read_sql_query(sql, getconn('capi2013'))
dfa['month'] = pd.to_datetime(dfa['date']).dt.month
dfa['year'] =  pd.to_datetime(dfa['date']).dt.year
dfa['amount'] =  dfa.apply(get_amount, axis=1) # return ("{:,}".format(number)) 
dfa

item      brand             model  qty     price        date  \
0              AC  PANASONIC           UC12PKY    1  19388.65  2013-04-01   
1          M/CARD       SONY             SR8N4    1    428.57  2013-04-01   
2          TABLET        HCL                Y3    1  11142.86  2013-04-01   
3          M/CARD       SONY             SR4N4    1    238.10  2013-04-01   
4      CALCULATOR      CASIO            MJ120D    1    371.43  2013-04-01   
5          TABLET        HCL                U1    1   5619.05  2013-04-01   
6        CELLULAR   VIDEOCON             V1539    1   1714.29  2013-04-01   
7        D/CAMERA      NIKON      COOLPIXS3500    1   6288.21  2013-04-01   
8        PSTATION       SONY          PSPE1004    1   6026.20  2013-04-01   
9        D/CAMERA       SONY         DSCHX200V    1  21397.38  2013-04-02   
10        BATTERY     PUREIT           GKB1500    1    393.01  2013-04-02   
11         M/CARD       SONY             SR4N4    1    257.14  2013-04-02   
12     CALCULATOR      CASIO            FX82MS    1    404.76  2013-04-02   
13       CELLULAR         LG              A290    1   2707.42  2013-04-02   
14       D/CAMERA    SAMSUNG            ECST72    1   5676.86  2013-04-02   
15       D/CAMERA       SONY          DSCWX150    1  11091.70  2013-04-02   
16     CALCULATOR      CASIO            FX82MS    1    404.76  2013-04-02   
17       DOMESTIC    PHILIPS            HP8310    1   1703.06  2013-04-02   
18       DOMESTIC    PHILIPS            HP8300    1   1135.37  2013-04-02   
19       CELLULAR      NOKIA              N100    1   1142.86  2013-04-02   
20       CELLULAR      NOKIA           ASHA308    1   4585.15  2013-04-02   
21         M/CARD    SANDISK         32GBULTRA    1   1619.05  2013-04-02   
22     CALCULATOR    CITIZEN           CT500JS    1    485.71  2013-04-02   
23      BINOCULAR  CELESTRON           10X25G2    1   1484.72  2013-04-02   
24       PENDRIVE   KINGSTON        DT100/16GB    1    619.05  2013-04-02   
25       V/CAMERA       SONY         HDRCX220E    1  17205.24  2013-04-02   
26       CELLULAR    SAMSUNG             i9082    1  18340.61  2013-04-02   
27         M/CARD       SONY            SF16N4    1    666.67  2013-04-02   
28           OVEN      HAIER        HDA1770EGT    1   3493.45  2013-04-02   
29        SCRATCH     MOLIFE             GUARD    1    190.48  2013-04-02   
...           ...        ...               ...  ...       ...         ...   
15361    CELLULAR      INTEX           CLOUDY1    1   3755.46  2014-03-31   
15362       PHONE  PANASONIC         KXTSC60SX    1    904.76  2014-03-31   
15363    CELLULAR    SAMSUNG             i9082    1  16157.21  2014-03-31   
15364    PENDRIVE    SANDISK              32GB    1   1095.24  2014-03-31   
15365    D/CAMERA      NIKON      COOLPIXS2800    1   4978.17  2014-03-31   
15366    D/CAMERA       SONY           DSCW830    1   6978.17  2014-03-31   
15367    D/CAMERA       SONY           DSCW810    1   6104.80  2014-03-31   
15368    D/CAMERA      NIKON        COOLPIXL29    1   4017.47  2014-03-31   
15369  CALCULATOR      CASIO            FX82MS    1    404.76  2014-03-31   
15370     BATTERY       SONY             NPBX1    1   1965.07  2014-03-31   
15371    CELLULAR    SAMSUNG             S7392    1   6637.55  2014-03-31   
15372  CALCULATOR      CASIO           FX991MS    1    690.48  2014-03-31   
15373    CELLULAR      INTEX           CLOUDY1    1   3842.79  2014-03-31   
15374         DVD       SONY          BDPS1100    1   5755.46  2014-03-31   
15375      M/CARD  KINGSTONE        8GBMICROSD    1    333.33  2014-03-31   
15376    PENDRIVE         HP         16GBV220W    1    666.67  2014-03-31   
15377    CELLULAR    SAMSUNG             S7582    1   8733.62  2014-03-31   
15378    CELLULAR    SAMSUNG             S5610    1   4628.82  2014-03-31   
15379    CELLULAR    SAMSUNG             S6312    1   5414.85  2014-03-31   
15380    DOMESTIC    PHILIPS            HP8100    1    676.86  2014-03-31   
15381    

In [54]:
import pandas as pd
import sqlanydb
from string import Template

mydf1 = pd.read_csv('mydata/test.txt')
sql = '''select item, brand, model, qty, price, "date"
    from bill_memo key join bill_memo_product
        key join inv_main
            key join product
                order by "date" '''
def getconn(db):
    return sqlanydb.connect(uid='dba', pwd='sql', eng='server', dbn=db, host='kushserver' )

def get_amount(row):
    return float(row['qty']) * float(row['price'])

pd.options.display.float_format = '{:,.0f}'.format

dfa = pd.read_sql_query(sql, getconn('capi2013'))
dfa['month'] = pd.to_datetime(dfa['date']).dt.month
dfa['year'] =  pd.to_datetime(dfa['date']).dt.year
dfa['amount'] =  dfa.apply(get_amount, axis=1)

dfb = pd.read_sql_query(sql, getconn('capi2014'))
dfb['month'] = pd.to_datetime(dfb['date']).dt.month
dfb['year'] =  pd.to_datetime(dfb['date']).dt.year
dfb['amount'] =  dfb.apply(get_amount, axis=1)

dfc = pd.read_sql_query(sql, getconn('capi2015'))
dfc['month'] = pd.to_datetime(dfc['date']).dt.month
dfc['year'] =  pd.to_datetime(dfc['date']).dt.year
dfc['amount'] =  dfc.apply(get_amount, axis=1)

df1 = pd.read_sql_query(sql, getconn('capi2016'))
df1['month'] = pd.to_datetime(df1['date']).dt.month
df1['year'] =  pd.to_datetime(df1['date']).dt.year
df1['amount'] =  df1.apply(get_amount, axis=1)

df2 = pd.read_sql_query(sql, getconn('capi2017'))
df2['month'] = pd.to_datetime(df2['date']).dt.strftime('%Y-%m')
df2['year'] =  pd.to_datetime(df2['date']).dt.year
df2['month'] = pd.to_datetime(df2['date']).dt.month
df2['amount'] =  df2.apply(get_amount, axis=1)

df3 = pd.read_sql_query(sql, getconn('capi2018'))
df3['month'] = pd.to_datetime(df3['date']).dt.strftime('%Y-%m')
df3['year'] =  pd.to_datetime(df3['date']).dt.year
df3['month'] = pd.to_datetime(df3['date']).dt.month
df3['amount'] =  df3.apply(get_amount, axis=1)

df4 = pd.read_sql_query(sql, getconn('capi2019'))
df4['month'] = pd.to_datetime(df4['date']).dt.strftime('%Y-%m')
df4['year'] =  pd.to_datetime(df4['date']).dt.year
df4['month'] = pd.to_datetime(df4['date']).dt.month
df4['amount'] =  df4.apply(get_amount, axis=1)

df = [dfa,dfb,dfc,df1,df2,df3,df4]
df = pd.concat(df)

p = df.pivot_table( index='item', columns='year', values='amount', aggfunc=sum, fill_value=0)
p.to_excel('c:/sales_amount.xlsx')
#p.plot(kind='line', x='item', y='year')
p.head(10000)


year            2013      2014      2015      2016      2017      2018    2019
item                                                                          
 FREEZER           0         0         0         0    18,253         0       0
 PURIFIER          0         0         0         0         0    15,848       0
*****              0         0         0     4,367   100,000   187,725 150,000
AC         1,426,453 2,591,607 1,560,146 1,744,716 1,579,128   712,805 636,179
ADAPTER        4,453         0         0     2,032     5,983         0     999
ADAPTOR        7,663     1,143     9,679     5,422         0         0   1,144
AIRFRYER           0         0         0    10,480         0         0       0
ALEXA              0         0         0         0         0    24,530       0
ANTIVIRUS     23,666    49,450    47,532    25,429    20,901    11,229       0
AUDIO        231,030   230,670   174,760    79,258    38,331     5,424  38,433
AVCABLE            0         0       306     1,135         0         0       0
B. COOLER          0    25,153         0         0         0         0       0
B.PACK             0         0    10,044         0         0         0       0
B/GRIP             0         0         0    16,769         0         0       0
BACKPACK           0         0         0     4,629     3,774       720       0
BAG           24,122     4,838     2,095         0       339         0  22,331
BALLHEAD           0         0         0         0         0         0   7,966
BAT/CHAG     120,348   183,845   140,936    55,361    67,771    55,210  33,263
BAT/GRIP           0         0         0         0    15,625    39,153       0
BATTARY            0         0         0         0         0         0   3,390
BATTERY      540,774   531,402   644,239   425,059   294,867   283,824 198,452
BDHTSYSTEM    44,978         0         0         0         0         0       0
BIGTV              0         0         0         0         0         0       0
BINOCULAR    369,738   229,607   249,694   296,017   323,377   287,026 166,871
BLENDER            0     2,052         0     1,397     2,726     3,737     932
BLUETOOTH          0         0     8,333     5,932         0    43,203  30,170
BODYCAP            0         0       262         0         0         0       0
BOTTLE INK         0         0         0         0         0     1,758       0
BP/MONITOR         0         0         0         0         0     3,304   1,653
BPMONITOR          0         0         0         0         0     3,768   1,780
...              ...       ...       ...       ...       ...       ...     ...
TORCH            649         0         0         0         0         0       0
TRIMER        16,332     5,790     4,087         0         0         0       0
TRIMMER            0         0         0     4,716    42,681   139,620  76,110
TRIPOD        65,231    54,306    75,746    32,157    61,528    90,787  36,737
TRIPOTHEAD         0         0         0     7,424     8,297         0       0
TROLLEY            0         0       786         0         0         0       0
TVTUNER            0     2,429     6,885     3,571         0         0       0
UPS            6,333     9,190    38,000     6,952     2,190     1,864       0
USB                0         0     1,305     2,095     1,021     1,639   1,954
USBCABLE           0         0     2,283       333       547    13,425  11,616
USBCHARGER     4,107    27,131   146,095    17,749    46,736    27,290   4,788
USBEX              0         0         0         0       540       212     170
USBHUB         1,571       419     1,809       714       476       517       0
UV                 0         0         0         0         0         0   7,424
UVFILTER     188,462   196,497   184,696   204,701   315,765   372,487 169,587
V/C                0    21,834    22,271         0         0         0       0
V/CAM              0         0         0         0   178,166    50,625       0
V/CAMARA           0         0         0         0

## Examples from Chapter 1

Read NSFG data into a Pandas DataFrame.

In [12]:
preg = nsfg.ReadFemPreg()
preg.head()

NameError: name 'nsfg' is not defined

Print the column names.

In [7]:
preg.columns

Index(['caseid', 'pregordr', 'howpreg_n', 'howpreg_p', 'moscurrp', 'nowprgdk',
       'pregend1', 'pregend2', 'nbrnaliv', 'multbrth',
       ...
       'laborfor_i', 'religion_i', 'metro_i', 'basewgt', 'adj_mod_basewgt',
       'finalwgt', 'secu_p', 'sest', 'cmintvw', 'totalwgt_lb'],
      dtype='object', length=244)

Select a single column name.

In [8]:
preg.columns[1]

'pregordr'

Select a column and check what type it is.

In [9]:
pregordr = preg['pregordr']
type(pregordr)

pandas.core.series.Series

Print a column.

In [10]:
pregordr

0        1
1        2
2        1
3        2
4        3
5        1
6        2
7        3
8        1
9        2
10       1
11       1
12       2
13       3
14       1
15       2
16       3
17       1
18       2
19       1
20       2
21       1
22       2
23       1
24       2
25       3
26       1
27       1
28       2
29       3
        ..
13563    2
13564    3
13565    1
13566    1
13567    1
13568    2
13569    1
13570    2
13571    3
13572    4
13573    1
13574    2
13575    1
13576    1
13577    2
13578    1
13579    2
13580    1
13581    2
13582    3
13583    1
13584    2
13585    1
13586    2
13587    3
13588    1
13589    2
13590    3
13591    4
13592    5
Name: pregordr, Length: 13593, dtype: int64

Select a single element from a column.

In [11]:
pregordr[0]

1

Select a slice from a column.

In [12]:
pregordr[2:5]

2    1
3    2
4    3
Name: pregordr, dtype: int64

Select a column using dot notation.

In [13]:
pregordr = preg.pregordr

Count the number of times each value occurs.

In [14]:
preg.outcome.value_counts().sort_index()

1    9148
2    1862
3     120
4    1921
5     190
6     352
Name: outcome, dtype: int64

Check the values of another variable.

In [15]:
preg.birthwgt_lb.value_counts().sort_index()

0.0        8
1.0       40
2.0       53
3.0       98
4.0      229
5.0      697
6.0     2223
7.0     3049
8.0     1889
9.0      623
10.0     132
11.0      26
12.0      10
13.0       3
14.0       3
15.0       1
Name: birthwgt_lb, dtype: int64

Make a dictionary that maps from each respondent's `caseid` to a list of indices into the pregnancy `DataFrame`.  Use it to select the pregnancy outcomes for a single respondent.

In [16]:
caseid = 10229
preg_map = nsfg.MakePregMap(preg)
indices = preg_map[caseid]
preg.outcome[indices].values

array([4, 4, 4, 4, 4, 4, 1], dtype=int64)

## Exercises

Select the `birthord` column, print the value counts, and compare to results published in the [codebook](http://www.icpsr.umich.edu/nsfg6/Controller?displayPage=labelDetails&fileCode=PREG&section=A&subSec=8016&srtLabel=611933)

In [17]:
# Solution goes here

We can also use `isnull` to count the number of nans.

In [18]:
preg.birthord.isnull().sum()

4445

Select the `prglngth` column, print the value counts, and compare to results published in the [codebook](http://www.icpsr.umich.edu/nsfg6/Controller?displayPage=labelDetails&fileCode=PREG&section=A&subSec=8016&srtLabel=611931)

In [19]:
# Solution goes here

To compute the mean of a column, you can invoke the `mean` method on a Series.  For example, here is the mean birthweight in pounds:

In [20]:
preg.totalwgt_lb.mean()

7.265628457623368

Create a new column named <tt>totalwgt_kg</tt> that contains birth weight in kilograms.  Compute its mean.  Remember that when you create a new column, you have to use dictionary syntax, not dot notation.

In [21]:
# Solution goes here

`nsfg.py` also provides `ReadFemResp`, which reads the female respondents file and returns a `DataFrame`:

In [22]:
resp = nsfg.ReadFemResp()

`DataFrame` provides a method `head` that displays the first five rows:

In [23]:
resp.head()

caseid  rscrinf  rdormres  rostscrn  rscreenhisp  rscreenrace  age_a  \
0    2298        1         5         5            1          5.0     27   
1    5012        1         5         1            5          5.0     42   
2   11586        1         5         1            5          5.0     43   
3    6794        5         5         4            1          5.0     15   
4     616        1         5         4            1          5.0     20   

   age_r  cmbirth  agescrn  ...  pubassis_i      basewgt  adj_mod_basewgt  \
0     27      902       27  ...           0  3247.916977      5123.759559   
1     42      718       42  ...           0  2335.279149      2846.799490   
2     43      708       43  ...           0  2335.279149      2846.799490   
3     15     1042       15  ...           0  3783.152221      5071.464231   
4     20      991       20  ...           0  5341.329968      6437.335772   

      finalwgt  secu_r  sest  cmintvw  cmlstyr  screentime   intvlngth  
0  5556.717241       2    18     1234     1222    18:26:36  110.492667  
1  4744.191350       2    18     1233     1221    16:30:59   64.294000  
2  4744.191350       2    18     1234     1222    18:19:09   75.149167  
3  5923.977368       2    18     1234     1222    15:54:43   28.642833  
4  7229.128072       2    18     1233     1221    14:19:44   69.502667  

[5 rows x 3087 columns]

Select the `age_r` column from `resp` and print the value counts.  How old are the youngest and oldest respondents?

In [24]:
# Solution goes here

We can use the `caseid` to match up rows from `resp` and `preg`.  For example, we can select the row from `resp` for `caseid` 2298 like this:

In [25]:
resp[resp.caseid==2298]

caseid  rscrinf  rdormres  rostscrn  rscreenhisp  rscreenrace  age_a  \
0    2298        1         5         5            1          5.0     27   

   age_r  cmbirth  agescrn  ...  pubassis_i      basewgt  adj_mod_basewgt  \
0     27      902       27  ...           0  3247.916977      5123.759559   

      finalwgt  secu_r  sest  cmintvw  cmlstyr  screentime   intvlngth  
0  5556.717241       2    18     1234     1222    18:26:36  110.492667  

[1 rows x 3087 columns]

And we can get the corresponding rows from `preg` like this:

In [26]:
preg[preg.caseid==2298]

caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1  \
2610    2298         1        NaN        NaN       NaN       NaN       6.0   
2611    2298         2        NaN        NaN       NaN       NaN       6.0   
2612    2298         3        NaN        NaN       NaN       NaN       6.0   
2613    2298         4        NaN        NaN       NaN       NaN       6.0   

      pregend2  nbrnaliv  multbrth  ...  laborfor_i  religion_i  metro_i  \
2610       NaN       1.0       NaN  ...           0           0        0   
2611       NaN       1.0       NaN  ...           0           0        0   
2612       NaN       1.0       NaN  ...           0           0        0   
2613       NaN       1.0       NaN  ...           0           0        0   

          basewgt  adj_mod_basewgt     finalwgt  secu_p  sest  cmintvw  \
2610  3247.916977      5123.759559  5556.717241       2    18      NaN   
2611  3247.916977      5123.759559  5556.717241       2    18      NaN   
2612  3247.916977      5123.759559  5556.717241       2    18      NaN   
2613  3247.916977      5123.759559  5556.717241       2    18      NaN   

      totalwgt_lb  
2610       6.8750  
2611       5.5000  
2612       4.1875  
2613       6.8750  

[4 rows x 244 columns]

How old is the respondent with `caseid` 1?

In [23]:
# Solution goes here

What are the pregnancy lengths for the respondent with `caseid` 2298?

In [24]:
# Solution goes here

What was the birthweight of the first baby born to the respondent with `caseid` 5012?

In [25]:
# Solution goes here